### PR TITLE
[SDK] Support local&cloud create run resume_from parameter

### PR DESCRIPTION
**Pull request by brynn-code**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.
This pull request introduces the feature to resume runs from existing runs in the PromptFlow SDK and CLI. The changes primarily revolve around the introduction of the `resume_from` parameter across various files and functions. The `resume_from` parameter allows users to create a run that resumes from an existing run. The new run reuses the succeeded line result of the original run and only runs the remaining/failed lines.

Key changes include:

Addition of `resume_from` functionality:

* [`src/promptflow/promptflow/_cli/_pf/_run.py`](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeR97-R100): Added `resume_from` parameter to the `add_run_create_common` function and the `create_run` function. The `create_run` function now checks for the presence of `resume_from` and calls the appropriate function accordingly. [[1]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeR97-R100) [[2]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeR113) [[3]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeL131-R144) [[4]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeL411-R423) [[5]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeL560-R575) [[6]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeR585) [[7]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeL581-R595) [[8]](diffhunk://#diff-0c9317bbb4cca7fd867af1f48509e5e17aa2a22fe09612467bd002ebfc6d1ebeL591-R640)

* [`src/promptflow/promptflow/_sdk/_pf_client.py`](diffhunk://#diff-c100045cd304dbaa66dd61b7087f78b813ce01e8566d0b823b293bb35755eccfL56-R56): Added `resume_from` parameter to the `run` function. The function now checks for the presence of `resume_from` and calls the appropriate function accordingly. [[1]](diffhunk://#diff-c100045cd304dbaa66dd61b7087f78b813ce01e8566d0b823b293bb35755eccfL56-R56) [[2]](diffhunk://#diff-c100045cd304dbaa66dd61b7087f78b813ce01e8566d0b823b293bb35755eccfR67) [[3]](diffhunk://#diff-c100045cd304dbaa66dd61b7087f78b813ce01e8566d0b823b293bb35755eccfR116-R144)

* [`src/promptflow/promptflow/_sdk/_run_functions.py`](diffhunk://#diff-3985b4097752d2257ce69e7105b3ece49a8a6238c22608a9e28a20218cd9a9bbR17-R22): Added `_resume_run` function to handle the creation of runs that resume from existing runs.

* [`src/promptflow/promptflow/_sdk/_submitter/run_submitter.py`](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cR42-R63): Added `resume` function to handle the resumption of runs. Modified `_run_bulk` and `_submit_bulk_run` functions to handle runs that resume from existing runs. [[1]](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cR42-R63) [[2]](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cL55-R80) [[3]](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cL83-R98) [[4]](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cR131-R134) [[5]](diffhunk://#diff-e3499cf713c66612d91723759de1c5076c22689b2ce2c89678f82de624b57c2cR143-R144)

Updates to documentation:

* [`docs/reference/pf-command-reference.md`](diffhunk://#diff-a7bd0e1379994523789706dd8bbbd61f2b4970ebda4e53a241a4c1ec0ef4a727R469): Updated the documentation for the `pf run create` command to include the `resume_from` parameter. [[1]](diffhunk://#diff-a7bd0e1379994523789706dd8bbbd61f2b4970ebda4e53a241a4c1ec0ef4a727R469) [[2]](diffhunk://#diff-a7bd0e1379994523789706dd8bbbd61f2b4970ebda4e53a241a4c1ec0ef4a727R498-R509)

* [`docs/reference/pfazure-command-reference.md`](diffhunk://#diff-eaed2450485fcb6f3d3fbe556d2c87ddf8e13f02bf3727df69dc146745075607R195): Updated the documentation for the `pfazure run create` command to include the `resume_from` parameter. [[1]](diffhunk://#diff-eaed2450485fcb6f3d3fbe556d2c87ddf8e13f02bf3727df69dc146745075607R195) [[2]](diffhunk://#diff-eaed2450485fcb6f3d3fbe556d2c87ddf8e13f02bf3727df69dc146745075607R245-R249)

* [`src/promptflow/CHANGELOG.md`](diffhunk://#diff-41ec3f7c4b5d4c0e670407d3c00a03a6966d7ebf617b1536473e33a12e2bc765R3-R13): Updated the changelog to include the new `resume_from` feature.

Other changes:

* [`src/promptflow/promptflow/_cli/_pf_azure/_run.py`](diffhunk://#diff-d1106ac9efec0530a00183e566fba9267ce3921980b402013fd4c84472a612fcR414): Updated the `dispatch_run_commands` function to handle runs that resume from existing runs.

* [`src/promptflow/promptflow/_sdk/_constants.py`](diffhunk://#diff-9ffef684cefef289f51e58b0aff1cd6cfa42363ace309d07f36dfe6559c5b84dR258): Added `RESUME_FROM` to the `FlowRunProperties` class.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_cli/_pf/_run.py [hunk 1]
```diff
@@ -406,6 +408,41 @@ def add_run_restore(subparsers):
     )
 
 
+def add_run_resume(subparsers):
```
Comment (brynn-code) [inline src/promptflow/promptflow/_cli/_pf/_run.py:411]: TODO: Finalize the customer facing api, add test and changelog

File: src/promptflow/promptflow/_sdk/entities/_run.py [hunk 2]
```diff
@@ -136,6 +136,7 @@ def __init__(
         self.tags = tags
         self.variant = variant
         self.run = run
+        self._resume_from = kwargs.get("resume_from", None)
```
Comment (brynn-code) [inline src/promptflow/promptflow/_sdk/entities/_run.py:139]: Shall we just change this to `self.resume_from`?

File: src/promptflow/promptflow/_cli/_pf/_run.py [hunk 3]
```diff
@@ -91,6 +94,10 @@ def add_run_create_common(subparsers, add_param_list, epilog: Optional[str] = No
     )
     add_param_name = lambda parser: parser.add_argument("-n", "--name", type=str, help="Name of the run.")  # noqa: E731
 
+    add_param_resume_from = lambda parser: parser.add_argument(  # noqa: E731
```
Comment (brynn-code) [inline src/promptflow/promptflow/_cli/_pf/_run.py:97]: Update changelog1.7.0 after release (also the doc)

Comment (crazygao) [reply src/promptflow/promptflow/_cli/_pf/_run.py:97]: lgtm after you update the changelog :D, pfs resume support in next patch?

Comment (brynn-code) [reply src/promptflow/promptflow/_cli/_pf/_run.py:97]: yes, will be a separate pr

File: src/promptflow/promptflow/_cli/_pf/_run.py [hunk 4]
```diff
@@ -578,48 +592,52 @@ def create_run(create_func: Callable, args):
...
+                "description": description,
+                "tags": tags,
+            }
+            return Run._load_from_source(source=run_source, params_override=processed_params)
         raise UserErrorException("To create a run, one of [file, flow, source] must be specified.")
-    run = create_func(run=run, stream=stream)
+
+    if resume_from:
+        if params_override:
+            logger.debug(f"Resume from specified, append params override {params_override} to run params.")
```
Comment (guming-learning) [inline src/promptflow/promptflow/_cli/_pf/_run.py:635]: resume_from

File: src/promptflow/promptflow/_sdk/_submitter/run_submitter.py [hunk 5]
```diff
@@ -39,6 +39,27 @@ def submit(self, run: Run, stream=False, **kwargs):
         self._run_bulk(run=run, stream=stream, **kwargs)
         return self.run_operations.get(name=run.name)
 
+    def resume(self, resume_from: str, **kwargs):
+        resume_from_run = self._ensure_required_run(resume_from)
+        run = resume_from_run._copy(resume_from=resume_from_run.name, **kwargs)
+        self._run_bulk(run=run, **kwargs)
+        return self.run_operations.get(name=run.name)
+
+    def _ensure_required_run(self, required_run: Union[str, Run]):
```
Comment (guming-learning) [inline src/promptflow/promptflow/_sdk/_submitter/run_submitter.py:48]: Please add comment to explain the behavior

File: src/promptflow/promptflow/_sdk/entities/_run.py [hunk 6]
```diff
@@ -701,3 +706,23 @@ def _load_from_source(cls, source: Union[str, Path], params_override: Optional[D
             end_time=datetime.datetime.fromisoformat(run_info["end_time"]),
             **kwargs,
         )
+
+    def _copy(self, **kwargs):
```
Comment (0mza987) [inline src/promptflow/promptflow/_sdk/entities/_run.py:710]: why do we need this? And we do, could change the name to be a more specific one, also improve the docstring about why we need it.

File: src/promptflow/promptflow/_sdk/_pf_client.py [hunk 7]
```diff
@@ -112,9 +113,18 @@ def run(
...
         :type tags: Dict[str, str]
+        :param resume_from: Create run resume from an existing run.
+        :type resume_from: str
         :return: Flow run info.
         :rtype: ~promptflow.entities.Run
         """
+        if resume_from:
+            unsupported = [flow, data, run, column_mapping, variant, connections, environment_variables]
+            if any(unsupported):
+                raise ValueError(f"'resume_from' is not supported with following parameters: {unsupported}. ")
```
Comment (guming-learning) [inline src/promptflow/promptflow/_sdk/_pf_client.py:124]: is not supported to be used with the following parameters

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 8]
```diff
@@ -107,6 +107,24 @@ def create_or_update(self, run: Run, **kwargs) -> Run:
         except RunExistsError:
             raise RunExistsError(f"Run {run.name!r} already exists.")
 
+    @monitor_operation(activity_name="pf.runs._resume", activity_type=ActivityType.PUBLICAPI)
+    def _resume(self, resume_from: str, **kwargs) -> Run:
```
Comment (guming-learning) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:111]: Seems the function name should be something like "_create_by_resuming_from"?
And should it be a public function?

File: src/promptflow/promptflow/_sdk/_constants.py [hunk 9]
```diff
@@ -251,6 +251,7 @@ class FlowRunProperties:
     # Experiment command node fields only
     COMMAND = "command"
     OUTPUTS = "outputs"
+    RESUME_FROM = "resume_from"
```
Comment (guming-learning) [inline src/promptflow/promptflow/_sdk/_constants.py:254]: For pfazure Run, are these properties stored in RunHistory property?

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py (hunk 8 above)
Comment (brynn-code) [reply src/promptflow/promptflow/_sdk/operations/_run_operations.py:111]: I'll take the name proposed, thanks.

File: src/promptflow/promptflow/_sdk/_constants.py (hunk 9 above)
Comment (brynn-code) [reply src/promptflow/promptflow/_sdk/_constants.py:254]: Offline synced.

**Review by 0mza987 — APPROVED** [review]

File: src/promptflow/promptflow/azure/_restclient/flow_service_caller.py [hunk 10]
```diff
@@ -675,6 +675,26 @@ def cancel_flow_run(
...
+        self,
+        subscription_id,  # type: str
+        resource_group_name,  # type: str
+        workspace_name,  # type: str
+        body=None,  # type: Optional["_models.ResumeBulkRunRequest"]
+        **kwargs,  # type: Any
+    ):
+        """Cancel a flow run."""
+        headers = self._get_headers()
+        return self.caller.bulk_runs.resume_bulk_run(
```
Comment (guming-learning) [inline src/promptflow/promptflow/azure/_restclient/flow_service_caller.py:689]: Seems `resume_bulk_run` is not implemented in _bulk_runs_operations.py yet?
I don't understand how does this function work now.

File: src/promptflow/promptflow/_sdk/_submitter/run_submitter.py [hunk 11]
```diff
@@ -39,6 +39,28 @@ def submit(self, run: Run, stream=False, **kwargs):
...
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_submitter/run_submitter.py:48]: maybe name this  _ensure_run_completed(), which will be more direct

File: src/promptflow/promptflow/_sdk/_submitter/run_submitter.py [hunk 12]
```diff
@@ -39,6 +39,28 @@ def submit(self, run: Run, stream=False, **kwargs):
...
+        elif not isinstance(required_run, Run):
+            error = TypeError(f"Referenced run must be a Run instance, got {type(required_run)}")
+            raise UserErrorException(message=str(error), error=error)
+        else:
+            # get the run again to make sure it's status is latest
+            required_run = self.run_operations.get(name=required_run.name)
+        if required_run.status != Status.Completed.value:
+            error = ValueError(f"Referenced run {required_run.name} is not completed, got status {required_run.status}")
+            raise UserErrorException(message=str(error), error=error)
+        required_run.outputs = self.run_operations._get_outputs(required_run)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_submitter/run_submitter.py:61]: why this?

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 13]
```diff
@@ -107,6 +107,24 @@ def create_or_update(self, run: Run, **kwargs) -> Run:
         except RunExistsError:
             raise RunExistsError(f"Run {run.name!r} already exists.")
 
+    @monitor_operation(activity_name="pf.runs._create_by_resume_from", activity_type=ActivityType.PUBLICAPI)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:110]: _create_by_resume_from -> resume

we can be simpler in telemetry

File: src/promptflow/promptflow/azure/_pf_client.py [hunk 14]
```diff
@@ -248,9 +249,21 @@ def run(
...
+        :param resume_from: Create run resume from an existing run.
+        :type resume_from: str
         :return: flow run info.
         :rtype: ~promptflow.entities.Run
         """
+        if resume_from:
+            unsupported = [flow, data, run, column_mapping, variant, connections, environment_variables]
+            if any(unsupported):
+                raise ValueError(
+                    f"'resume_from' is not supported to be used with the with following parameters: {unsupported}. "
```
[the remark this answers is not in the document]
Comment (brynn-code) [reply src/promptflow/promptflow/azure/_pf_client.py:261]: Good catch, will refine this.

File: src/promptflow/promptflow/azure/_restclient/flow_service_caller.py (hunk 10 above)
Comment (brynn-code) [reply src/promptflow/promptflow/azure/_restclient/flow_service_caller.py:689]: Offline synced. that resume_bulk_run is restclient code generated by autorest.

File: src/promptflow/promptflow/_sdk/_submitter/run_submitter.py (hunk 12 above)
Comment (brynn-code) [reply src/promptflow/promptflow/_sdk/_submitter/run_submitter.py:61]: (I don't have enough context about this lol

**Review by wangchao1230 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**